File: references/betterlectio/ios/BetterLectio/GradeParser.swift
//
//  GradeParser.swift
//  BetterLectio
//

import Foundation
import SwiftSoup

/// Parser for grades and report cards from Lectio.
enum GradeParser {
    
    // MARK: - Parse Grades Report

    /// Parses the Lectio grades report page (`grades/grade_report.aspx`) into structured data.
    static func parseGradesReport(from html: String) throws -> GradesReport {
        let doc = try SwiftSoup.parse(html)

        let writtenAlertText = try doc
            .select("#s_m_Content_Content_karakterView_WrittenProtokolBlockLit")
            .first()?
            .text()
            .trimmingCharacters(in: .whitespacesAndNewlines)

        let oralAlertText = try doc
            .select("#s_m_Content_Content_karakterView_OralProtokolBlockLit")
            .first()?
            .text()
            .trimmingCharacters(in: .whitespacesAndNewlines)

        let blockedWrittenProtocolTerm = extractBlockedTerm(fromAlert: writtenAlertText)
        let blockedOralProtocolTerm = extractBlockedTerm(fromAlert: oralAlertText)

        var grades: [GradeEntry] = []
        var columns: [GradeColumn] = []

        if let gradesTable = try doc.select("#s_m_Content_Content_karakterView_KarakterGV").first() {
            let rows = try gradesTable.select("tr")

            if let headerRow = rows.first(where: { row in
                guard let count = try? desktopHeaderCells(row).count else { return false }
                return count > 2
            }) {
                let headerCells = try desktopHeaderCells(headerRow)
                var usedKeys: Set<String> = []
                for header in headerCells.dropFirst(2) {
                    let label = try header.text().collapsingWhitespace
                    guard !label.isEmpty else { continue }
                    let baseKey = canonicalColumnKey(label)
                    var key = baseKey
                    var suffix = 2
                    while usedKeys.contains(key) {
                        key = "\(baseKey)-\(suffix)"
                        suffix += 1
                    }
                    usedKeys.insert(key)
                    columns.append(GradeColumn(key: key, label: label))
                }
            }

            for row in rows {
                // Parse desktop cells only to avoid duplicate mobile table markup.
                let cells = try desktopDataCells(row)
                guard !columns.isEmpty, cells.count >= 2 + columns.count else { continue }

                let holdSpan = try cells[0].select("span[data-lectiocontextcard]").first()
                let hold = try holdSpan?.text().trimmingCharacters(in: .whitespacesAndNewlines)
                    ?? cells[0].text().trimmingCharacters(in: .whitespacesAndNewlines)
                let holdElementId = try holdSpan?.attr("data-lectiocontextcard")

                let subject = try cells[1].text().trimmingCharacters(in: .whitespacesAndNewlines)

                var gradeCells: [String: GradeCellValue] = [:]
                for (offset, column) in columns.enumerated() {
                    if let value = try parseGradeCell(cells[offset + 2]) {
                        gradeCells[column.key] = value
                    }
                }

                let idBase = [hold, subject].joined(separator: "|")
                let id = idBase.isEmpty ? UUID().uuidString : idBase

                grades.append(GradeEntry(
                    id: id,
                    hold: hold,
                    holdElementId: holdElementId,
                    subject: subject,
                    grades: gradeCells
                ))
            }
        }

        var notes: [GradeNoteEntry] = []

        if let notesTable = try doc.select("#s_m_Content_Content_karakterView_KarakterNoterGrid").first() {
            let rows = try notesTable.select("tr")

            for row in rows {
                let cells = try row.select("td.OnlyDesktop")
                guard cells.count >= 5 else { continue }

                let hold = try cells[0].text().trimmingCharacters(in: .whitespacesAndNewlines)
                let gradeType = try cells[1].text().trimmingCharacters(in: .whitespacesAndNewlines)
                let grade = try cells[2].text().trimmingCharacters(in: .whitespacesAndNewlines)
                let insertedAt = try cells[3].text().trimmingCharacters(in: .whitespacesAndNewlines)
                let note = try cells[4].text().trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty

                let idBase = [hold, gradeType, insertedAt].joined(separator: "|")
                let id = idBase.isEmpty ? UUID().uuidString : idBase

                notes.append(GradeNoteEntry(
                    id: id,
                    hold: hold,
                    gradeType: gradeType,
                    grade: grade,
                    insertedAt: insertedAt,
                    note: note
                ))
            }
        }

        print("📊 Parsed grades report: \(grades.count) grade rows, \(notes.count) note rows")

        return GradesReport(
            blockedWrittenProtocolTerm: blockedWrittenProtocolTerm,
            blockedOralProtocolTerm: blockedOralProtocolTerm,
            columns: columns,
            grades: grades,
            notes: notes
        )
    }

    /// Stable keys let selection survive refreshes while the original label is
    /// retained for display. Unknown future columns deliberately remain usable.
    static func canonicalColumnKey(_ raw: String) -> String {
        let value = raw.collapsingWhitespace.lowercased()
        if value.range(of: #"^1\.?\s*standpunkt"#, options: .regularExpression) != nil { return "1.standpunkt" }
        if value.range(of: #"^2\.?\s*standpunkt"#, options: .regularExpression) != nil { return "2.standpunkt" }
        if value.range(of: #"^3\.?\s*standpunkt"#, options: .regularExpression) != nil { return "3.standpunkt" }
        if value.contains("afsluttende") { return "afsluttende" }
        if value.contains("intern") { return "intern prøve" }
        if value.contains("eksamen") { return "eksamenskarakter" }
        if value.hasPrefix("årskarakter") || value.hasPrefix("aarskarakter") { return "årskarakter" }

        let slug = value
            .replacingOccurrences(
                of: "[^a-z0-9æøå]+",
                with: "-",
                options: .regularExpression
            )
            .trimmingCharacters(in: CharacterSet(charactersIn: "-"))
        return slug.isEmpty ? raw : slug
    }

    private static func desktopHeaderCells(_ row: Element) throws -> Elements {
        let desktop = try row.select("th.OnlyDesktop, td.OnlyDesktop")
        if desktop.count > 0 { return desktop }
        return try row.select("th")
    }

    private static func desktopDataCells(_ row: Element) throws -> Elements {
        let desktop = try row.select("td.OnlyDesktop")
        if desktop.count > 0 { return desktop }
        return try row.select("td:not(.OnlyMobile)")
    }

    private static func parseGradeCell(_ cell: Element) throws -> GradeCellValue? {
        let valueText = try cell.text().trimmingCharacters(in: .whitespacesAndNewlines)
        guard !valueText.isEmpty else { return nil }

        // Metadata usually lives on an inner div title attribute, e.g.:
        // XPRSFag: 4682C Musik\nKilde: Karakter\nVægt: 1
        let title = try cell.select("div[title]").first()?.attr("title")

        let xprsSubject = extractMetadataValue(prefix: "XPRSFag:", from: title)
        let source = extractMetadataValue(prefix: "Kilde:", from: title)
        let weightRaw = extractMetadataValue(prefix: "Vægt:", from: title)
        let weight = weightRaw
            .map { $0.replacingOccurrences(of: ",", with: ".") }
            .flatMap(Double.init)

        return GradeCellValue(
            value: valueText,
            xprsSubject: xprsSubject,
            source: source,
            weight: weight
        )
    }

    private static func extractMetadataValue(prefix: String, from title: String?) -> String? {
        guard let title, !title.isEmpty else { return nil }
        guard let line = title.components(separatedBy: .newlines)
            .lazy
            .map({ $0.trimmingCharacters(in: .whitespacesAndNewlines) })
            .first(where: { $0.hasPrefix(prefix) })
        else { return nil }
        let value = line.replacingOccurrences(of: prefix, with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return value.isEmpty ? nil : value
    }

    private static func extractBlockedTerm(fromAlert text: String?) -> String? {
        guard let text, !text.isEmpty else { return nil }
        guard let range = text.range(of: #"terminen\s+\"([^\"]+)\""#, options: .regularExpression) else {
            return nil
        }

        let matched = String(text[range])
        let stripped = matched
            .replacingOccurrences(of: "terminen", with: "")
            .replacingOccurrences(of: "\"", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return stripped.isEmpty ? nil : stripped
    }
}

private extension String {
    var collapsingWhitespace: String {
        split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
    }
}

File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/grades/GradeParser.kt
package dk.betterlectio.android.feature.grades

import org.jsoup.Jsoup
import org.jsoup.nodes.Element

/**
 * Grades from `grade_report.aspx` / KarakterGV.
 *
 * Column identity and order are taken from the live header row — Lectio varies
 * which columns exist per school/term (e.g. "Afsluttende års-/standpunktskarakter").
 * A fixed column list silently shifts values and can swap årskarakter/eksamen.
 * Mirrors extension `parseKaraktererFromDOM` / `canonicalColumnKey`.
 */
object GradeParser {

    fun parse(html: String): GradesReport {
        val doc = Jsoup.parse(html)
        val columns = mutableListOf<GradeColumn>()
        val grades = mutableListOf<GradeRow>()

        val table = doc.getElementById("s_m_Content_Content_karakterView_KarakterGV")
            ?: doc.selectFirst("table[id*=KarakterGV], table[id*=Karakter]")

        if (table != null) {
            val rows = table.select("tr")
            if (rows.isNotEmpty()) {
                // Header: first two desktop <th> are Hold + Fag; rest are grade columns.
                val headerCells = desktopHeaderCells(rows[0])
                for (h in 2 until headerCells.size) {
                    val label = headerCells[h].text().replace(Regex("\\s+"), " ").trim()
                    if (label.isEmpty()) continue
                    columns += GradeColumn(key = canonicalColumnKey(label), label = label)
                }

                for (i in 1 until rows.size) {
                    val cells = desktopCells(rows[i])
                    if (cells.size < 2 + columns.size) continue

                    val teamCell = cells[0]
                    val teamSpan = teamCell.selectFirst("[data-lectiocontextcard], [data-lectioContextCard]")
                    val team = teamSpan?.text()?.trim() ?: teamCell.text().trim()
                    val teamId = teamSpan?.attr("data-lectiocontextcard")
                        ?.ifBlank { null }
                        ?: teamSpan?.attr("data-lectioContextCard")?.ifBlank { null }
                    val subject = cells[1].text().trim()

                    val gradeMap = linkedMapOf<String, GradeCellValue>()
                    for (c in columns.indices) {
                        val cell = cells.getOrNull(c + 2) ?: continue
                        val parsed = parseGradeCell(cell) ?: continue
                        gradeMap[columns[c].key] = parsed
                    }

                    grades += GradeRow(
                        team = team,
                        subject = subject.ifBlank { team },
                        teamId = teamId,
                        grades = gradeMap,
                    )
                }
            }
        }

        return GradesReport(
            columns = columns,
            grades = grades,
            notes = parseNotesStructured(doc),
            alerts = parseAlerts(doc),
        )
    }

    fun parseNotes(html: String): List<GradeNoteEntry> {
        val doc = Jsoup.parse(html)
        return parseNotesStructured(doc)
    }

    /**
     * Maps a Lectio header label to a stable key so the rest of the UI can
     * recognise well-known columns. Order of checks matters: "afsluttende" and
     * "eksamen" labels both contain "års"-ish wording.
     */
    fun canonicalColumnKey(raw: String): String {
        val t = raw.lowercase().replace(Regex("\\s+"), " ").trim()
        if (Regex("^1\\.?\\s*standpunkt").containsMatchIn(t)) return "1.standpunkt"
        if (Regex("^2\\.?\\s*standpunkt").containsMatchIn(t)) return "2.standpunkt"
        if (Regex("^3\\.?\\s*standpunkt").containsMatchIn(t)) return "3.standpunkt"
        if (t.contains("afsluttende")) return "afsluttende"
        if (t.contains("intern")) return "intern prøve"
        if (t.contains("eksamen")) return "eksamenskarakter"
        if (t.startsWith("årskarakter") || t.startsWith("aarskarakter")) return "årskarakter"
        val slug = t.replace(Regex("[^a-z0-9æøå]+"), "-").trim('-')
        return slug.ifEmpty { raw }
    }

    private fun parseAlerts(doc: org.jsoup.nodes.Document): List<String> {
        val ids = listOf(
            "s_m_Content_Content_karakterView_WrittenProtokolBlockLit",
            "s_m_Content_Content_karakterView_OralProtokolBlockLit",
        )
        return ids.mapNotNull { id ->
            doc.getElementById(id)?.text()?.trim()?.takeIf { it.isNotEmpty() }
        }
    }

    private fun parseNotesStructured(doc: org.jsoup.nodes.Document): List<GradeNoteEntry> {
        val grid = doc.getElementById("s_m_Content_Content_karakterView_KarakterNoterGrid")
            ?: doc.selectFirst("table[id*=KarakterNoter], table[id*=NoterGrid]")
            ?: return emptyList()

        return grid.select("tr").drop(1).mapNotNull { row ->
            val cells = desktopCells(row)
            if (cells.size < 4) return@mapNotNull null

            val hold = cells[0].text().trim()
            val gradeType = cells[1].text().replace(Regex("\\s+"), " ").trim()
            val grade = cells[2].text().trim()
            val insertedAt = cells[3].text().trim()
            // Note is usually the last desktop cell (may include wrap OnlyDesktop).
            val noteCell = cells.lastOrNull()
            val noteText = noteCell?.let { cell ->
                // Prefer last cell when there are ≥5 desktop cells; if only 4, no note body.
                if (cells.size >= 5) cell.wholeText().trim().ifEmpty { cell.text().trim() } else null
            }?.takeIf { it.isNotEmpty() && it != grade && it != insertedAt && it != hold && it != gradeType }

            if (hold.isEmpty() && gradeType.isEmpty()) return@mapNotNull null

            GradeNoteEntry(
                hold = hold,
                gradeType = gradeType,
                grade = grade,
                insertedAt = insertedAt,
                note = noteText,
            )
        }.distinct().take(100)
    }

    private fun parseGradeCell(cell: Element): GradeCellValue? {
        val titled = cell.selectFirst("div[title], span[title]")
        val valueEl = titled ?: cell.children().firstOrNull() ?: cell
        val text = valueEl.text().trim()
        if (text.isEmpty() || text == "--" || text == "–" || text == "—") return null

        val title = titled?.attr("title")
            ?.ifBlank { null }
            ?: cell.attr("title").ifBlank { null }

        return GradeCellValue(
            value = text,
            weight = extractMetadataValue(title, "Vægt:", "Weight:")
                ?.replace(",", ".")
                ?.toDoubleOrNull(),
            xprsSubject = extractMetadataValue(title, "XPRSFag:"),
            source = extractMetadataValue(title, "Kilde:", "Source:"),
        )
    }

    private fun extractMetadataValue(title: String?, vararg prefixes: String): String? {
        if (title.isNullOrBlank()) return null
        for (rawLine in title.lineSequence()) {
            val line = rawLine.trim()
            for (prefix in prefixes) {
                if (line.startsWith(prefix, ignoreCase = true)) {
                    val value = line.substring(prefix.length).trim()
                    if (value.isNotEmpty()) return value
                } else if (line.contains(prefix, ignoreCase = true) &&
                    (prefix.startsWith("Vægt") || prefix.startsWith("Weight"))
                ) {
                    // "Vægt: 1" mid-line fallback used by some titles
                    val after = line.substringAfter(':', missingDelimiterValue = "")
                        .trim()
                        .replace(",", ".")
                    if (after.isNotEmpty()) return after
                }
            }
        }
        return null
    }

    private fun desktopHeaderCells(row: Element): List<Element> {
        val desktop = row.select("th.OnlyDesktop")
        if (desktop.isNotEmpty()) return desktop
        val nonMobile = row.select("th:not(.OnlyMobile)")
        if (nonMobile.isNotEmpty()) return nonMobile
        return row.select("th")
    }

    private fun desktopCells(row: Element): List<Element> {
        // Prefer OnlyDesktop; include wrap.OnlyDesktop for notes grid.
        val desktop = row.select("td.OnlyDesktop, td.wrap.OnlyDesktop")
        if (desktop.isNotEmpty()) return desktop
        val nonMobile = row.select("td:not(.OnlyMobile)")
        if (nonMobile.isNotEmpty()) return nonMobile
        return row.select("td")
    }
}
